Docs: HogQL expressions (#6597)
PostHog/posthog.com: contents/docs/hogql/expressions.md

---
title: HogQL expressions
sidebar: Docs
showTitle: true
availability:
    free: full
    selfServe: full
    enterprise: full
---

HogQL expressions enable you to directly access, modify, and aggregate data in PostHog using SQL. You can use them nearly everywhere where event filters exist, including:

> - Dashboards

Reviewer: ```suggestion
- [Dashboards](/docs/product-analytics/dashboards)
```

Reviewer: Suggestion, although maybe we dont need to include these hyperlinks

- Data series
- Breakdowns

> - Funnels

Reviewer: ```suggestion
- [Funnels](/docs/product-analytics/funnels)
```

- Event explorer

![HogQL trends breakdown filter](../../images/features/hogql/trends-breakdown.png)

HogQL expressions can access data like:

> - event `properties`
> - person `properties`

Reviewer: Maybe dont need this hyperlink 

```suggestion
- event properties
- [person properties](/docs/product-analytics/user-properties)
```

- `event`
- `elements_chain`
- `timestamp`
- `distinct_id`
- `person_id`

> They then use SQL functions to access, filter, modify, or aggregate the data. A full list of SQL functions are found in [supported ClickHouse functions](/docs/hogql/clickhouse-functions) and [supported aggregations](/docs/hogql/aggregations).

Reviewer: ```suggestion
They then use SQL functions to access, filter, modify, or aggregate the data. A full list of SQL functions are found in our [supported ClickHouse functions](/docs/hogql/clickhouse-functions) and [supported aggregations](/docs/hogql/aggregations) docs.
```

> > **Tip:** If you're having trouble getting results from you expression, try testing with a different visualization (trends table often works best) or break down your expression into pieces and test each one.

Reviewer: >  try testing with a different visualization (trends table often works best)

Can you elaborate why?

Author: Basically, you can see all the values returned. Added that

> > **Tip:** If you're having trouble getting results from you expression, try testing with a different visualization (trends table often works best as it shows all values returned) or break down your expression into pieces and test each one.

Reviewer: ```suggestion
> **Tip:** If you're having trouble getting results from your expression, try debugging by testing with a different visualization (trends table often works best as it shows all values returned) or break down your expression into pieces and test each one.
```

### Dates

- `dateDiff('unit', startdate, enddate)`: Returns the count in `unit` between `startdate` and `enddate`.
- `toDayOfWeek`, `toHour`, `toMinute`: Converts date number of day of week (1-7), hour in 24-hour time (0-23), and minute in hour (0-59).
- `now()`, `today()`, `yesterday()`: Returns the current time, date, or yesterday’s date respectively.
- `interval`: A length of time for use in arithmetic operations with other dates and times.

## Use cases

- Checking if a property or autocapture element chain contains a specific value or any of an array of values using `in` or `match`.
- Modifying the display string in the visualization by extracting or concatenating properties using `concat()`, `+`, `extract()`, or `replaceOne`.
- Grouping or binning events based on properties using `if()`, `multiIf()`.
- Accessing nested properties such as `properties.$set.$geoip_city_name`.
- Filtering for events that happened in the last X minutes or hours with `dateDiff()`, `now()`, and `interval`.
- Creating percentages by calculating the sum of one property over the sum of all related properties with `sum()`, `/`, `+`, and `*`.
- Getting unique values with `uniq()`.

> - Binning events based on time of day, week, and month with `toHour`, `toDayOfWeek`, `toStartOfWeek`, `toMonth`.

Reviewer: I think for each of these it would be good to include an example with a code snippet!
